src/Build/BackEnd/Components/SdkResolution/NuGet/NuGetSdkResolver.cs

```diff
@@ -0,0 +1,202 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Build.Framework;
+using NuGet.Commands;
+using NuGet.Configuration;
+using NuGet.LibraryModel;
+using NuGet.Packaging;
+using NuGet.Versioning;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+using SdkResolverContextBase = Microsoft.Build.Framework.SdkResolverContext;
+using SdkResultBase = Microsoft.Build.Framework.SdkResult;
+using SdkResultFactoryBase = Microsoft.Build.Framework.SdkResultFactory;
+
+namespace Microsoft.Build.BackEnd.SdkResolution.NuGet
+{
+    /// <summary>
+    /// Represents a NuGet-based SDK resolver.  It is very important that this class does not reference any NuGet assemblies
+    /// directly as an optimization to avoid loading them unless they are needed.  The current implementation only loads
+    /// Newtonsoft.Json if a global.json is found and it contains the msbuild-sdks section and a few NuGet assemblies to parse
+    /// a version.  The remaining NuGet assemblies are then loaded to do a restore.
+    /// </summary>
+    internal sealed class NuGetSdkResolver : NuGetSdkResolverBase
+    {
+        public override string Name => nameof(NuGetSdkResolver);
+
+        public override int Priority => 2500;
```

Comment:
To be explicit here for everyone reviewing, the `DefaultSdkResolver` is 10,000 and `DotNetMSBuildSdkResolver` is 5,000. So at 2,500 `NuGetSdkResolver` will run before any others (that we ship).